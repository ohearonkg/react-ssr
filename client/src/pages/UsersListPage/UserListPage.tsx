import { AxiosInstance } from "axios";
import * as React from "react";
import { connect } from "react-redux";
import { AnyAction, Dispatch, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import { fetchUsersFunction } from "../../actions/Users/users";
import { IUser } from "../../models/User";
import { ApplicationState } from "../../reducers";

interface IUserListPage {
  /**
   * Users to be rendered
   */
  users: IUser[];

  /**
   * Function to fetch users
   */
  fetchUsers: () => void;
}

class UserListPage extends React.Component<IUserListPage, {}> {
  public componentDidMount() {
    this.props.fetchUsers();
  }

  public renderUsers = () => {
    const { users } = this.props;
    return users.map((user: IUser) => <li key={user.id}> {user.name} </li>);
  };

  public render() {
    return (
      <div>
        <h1> User List </h1>
        <ul> {this.renderUsers()} </ul>
      </div>
    );
  }
}

const mapStateToProps = (state: ApplicationState) => {
  return {
    users: state.users.users
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    // @ts-ignore
    fetchUsers: () => dispatch(fetchUsersFunction())
  };
};

/**
 * Thunk called to load data on the
 * server side
 *
 * Arguments passed in include the axios instance
 * which is different between client and server. The
 * client side requests MUST be proxied.
 */
const loadData = (
  store: Store<ApplicationState, AnyAction> & {
    dispatch: ThunkDispatch<{}, AxiosInstance, AnyAction>;
  }
) => store.dispatch(fetchUsersFunction());

export { UserListPage };

export default {
  component: connect(
    mapStateToProps,
    mapDispatchToProps
  )(UserListPage),
  loadData
};
