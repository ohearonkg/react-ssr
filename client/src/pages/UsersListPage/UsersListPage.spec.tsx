import { shallow, ShallowWrapper } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { UserListPage } from "./UserListPage";

describe("Users List Page", () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <UserListPage
        users={[
          {
            id: 1,
            name: "a"
          },
          {
            id: 2,
            name: "b"
          },
          {
            id: 3,
            name: "c"
          }
        ]}
      />
    );
  });

  /**
   * Snapshot
   */
  it("Should match its snapshot", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  /**
   * Showing list of users
   */
  it("Should render a list of users passed to its users prop", () => {
    expect(wrapper.find("li")).toHaveLength(3);
  });
});
