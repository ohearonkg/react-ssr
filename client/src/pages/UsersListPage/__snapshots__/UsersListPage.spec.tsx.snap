// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users List Page Should match its snapshot 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
       
      React SSR - Users (3)
       
    </title>
    <meta
      content="React SSR - Users"
      name="og:title"
    />
  </HelmetWrapper>
  <h1>
     User List 
  </h1>
  <ul>
     
    <li
      key="1"
    >
       
      a
       
    </li>
    <li
      key="2"
    >
       
      b
       
    </li>
    <li
      key="3"
    >
       
      c
       
    </li>
     
  </ul>
</div>
`;
